### PR TITLE
Add new ROS2 dynamic SH

### PR DESCRIPTION
**Pull request by lauramg15**
**TODO:**
* [ ]  Update README
* [ ] Update documentation
* [x] Adapt tests to execute them both in static and dynamic mode

### CODE REVIEW

File: ros2/CMakeLists.txt [hunk 1]
```diff
@@ -64,159 +52,18 @@ if(BUILD_LIBRARY)
...
-###################################################################################
-# Integration Service ROS 2 SystemHandle API Reference
-###################################################################################
-if(BUILD_API_REFERENCE)
-    if(NOT BUILD_LIBRARY)
-        find_file(IS_DOXYGEN_CONFIG_FILE doxygen-config.in PATHS ${CMAKE_INSTALL_PREFIX}/../is-core)
+    if("${IS_ROS2_SH_MODE_LOWERCASE}" STREQUAL "static")
+        add_subdirectory(static)
+    elseif("${IS_ROS2_SH_MODE_LOWERCASE}" STREQUAL "dynamic")
+        add_subdirectory(dynamic)
```
Comment (jamoralp) [inline ros2/CMakeLists.txt]:
```suggestion
        add_subdirectory(dynamic)
    else
        message(WARNING "Invalid mode selected. Please choose between 'static' or 'dynamic'")
```

File: ros2/dynamic/CMakeLists.txt.user [hunk 2]
```diff
@@ -0,0 +1,371 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jamoralp) [inline ros2/dynamic/CMakeLists.txt.user]: This file should not be updated

File: ros2/dynamic/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+#     # Install doxygen generated XML files
+#     install(DIRECTORY ${PROJECT_BINARY_DIR}/doxygen/xml
+#         DESTINATION doxygen
+#         COMPONENT doxygen-${PROJECT_NAME})
+#     set(CPACK_COMPONENT_EXAMPLES_DISPLAY_NAME "ROS2 System Handle doxygen")
+#     set(CPACK_COMPONENT_EXAMPLES_DESCRIPTION
+#         "eProsima ROS2 System Handle doxygen documentation")
+#     set(CPACK_COMPONENTS_ALL ${CPACK_COMPONENTS_ALL} doxygen-${PROJECT_NAME})
+# endif()
```
Comment (jamoralp) [inline ros2/dynamic/CMakeLists.txt]: Delete this

File: ros2/dynamic/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,224 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
```
Comment (jamoralp) [inline ros2/dynamic/CMakeLists.txt:1]: If this whole CMakeLists is conditioned to `BUILD_LIBRARY`, we could remove the `if(BUILD_LIBRARY)` clauses here and add them in the top CMakeLists as a condition to `add_subdirectory`

File: ros2/dynamic/include/is/sh/ros2/Conversion.hpp [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef _IS_SH_FASTDDS__INTERNAL__CONVERSION_HPP_
+#define _IS_SH_FASTDDS__INTERNAL__CONVERSION_HPP_
```
Comment (jamoralp) [inline ros2/dynamic/include/is/sh/ros2/Conversion.hpp]:
```suggestion
#ifndef _IS_SH_ROS2_DYNAMIC__INTERNAL__CONVERSION_HPP_
#define _IS_SH_ROS2_DYNAMIC__INTERNAL__CONVERSION_HPP_
```

File: ros2/dynamic/include/is/sh/ros2/Conversion.hpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+    static utils::Logger logger_;
+};
+
+} //  namespace ros2
+} //  namespace sh
+} //  namespace is
+} //  namespace eprosima
+
+#endif //  _IS_SH_FASTDDS__INTERNAL__CONVERSION_HPP_
```
Comment (jamoralp) [inline ros2/dynamic/include/is/sh/ros2/Conversion.hpp]:
```suggestion
#endif //  _IS_SH_ROS2_DYNAMIC__INTERNAL__CONVERSION_HPP_
```

File: ros2/dynamic/include/is/sh/ros2/Participant.hpp [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#ifndef _IS_SH_FASTDDS__INTERNAL__PARTICIPANT_HPP_
+#define _IS_SH_FASTDDS__INTERNAL__PARTICIPANT_HPP_
```
Comment (jamoralp) [inline ros2/dynamic/include/is/sh/ros2/Participant.hpp]:
```suggestion
#ifndef _IS_SH_ROS2_DYNAMIC__INTERNAL__PARTICIPANT_HPP_
#define _IS_SH_ROS2_DYNAMIC__INTERNAL__PARTICIPANT_HPP_
```

File: ros2/dynamic/include/is/sh/ros2/Publisher.hpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef _IS_SH_FASTDDS__INTERNAL__PUBLISHER_HPP_
+#define _IS_SH_FASTDDS__INTERNAL__PUBLISHER_HPP_
```
Comment (jamoralp) [inline ros2/dynamic/include/is/sh/ros2/Publisher.hpp]: Replace all occurrences. I will ommit the rest of them

File: ros2/dynamic/include/is/sh/ros2/Subscriber.hpp [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+/**
+ * @brief Forward declaration.
+ */
+class Participant;
+
+/**
+ * @class Subscriber
+ *        This class represents a <a href="https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/subscriber/subscriber.html">
+ *        Fast DDS Subscriber</a> within the *Integration Service* framework.
```
Comment (jamoralp) [inline ros2/dynamic/include/is/sh/ros2/Subscriber.hpp]:
```suggestion
 *        ROS 2 Subscriber</a> within the *Integration Service* framework.

```

File: ros2/dynamic/src/Publisher.cpp [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+    // Create DDS datawriter
+    ::fastdds::dds::DataWriterQos datawriter_qos = ::fastdds::dds::DATAWRITER_QOS_DEFAULT;
+    if (config["service_instance_name"])
+    {
+        fastrtps::rtps::Property instance_property;
+        instance_property.name("dds.rpc.service_instance_name");
+        instance_property.value(config["service_instance_name"].as<std::string>());
+        datawriter_qos.properties().properties().emplace_back(std::move(instance_property));
+    }
```
Comment (jamoralp) [inline ros2/dynamic/src/Publisher.cpp]: This is probably not needed

File: ros2/dynamic/src/Publisher.cpp [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+bool Publisher::publish(
+        const ::xtypes::DynamicData& message)
+{
+    std::unique_lock<std::mutex> lock(data_mtx_);
+
+    logger_ << utils::Logger::Level::INFO
+            << "Sending message from Integration Service to DDS for topic '" << topic_name_ << "': "
```
Comment (jamoralp) [inline ros2/dynamic/src/Publisher.cpp]:
```suggestion
            << "Sending message from Integration Service to ROS 2 for topic '" << topic_name_ << "': "

```

File: ros2/dynamic/src/Publisher.cpp [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+
+    bool success = Conversion::xtypes_to_fastdds(message, dynamic_data_);
+    if (success)
+    {
+        success = dds_datawriter_->write(static_cast<void*>(dynamic_data_));
+    }
+    else
+    {
+        logger_ << utils::Logger::Level::ERROR
+                << "Failed to convert message from Integration Service to DDS for topic '"
```
Comment (jamoralp) [inline ros2/dynamic/src/Publisher.cpp]:
```suggestion
                << "Failed to convert message from Integration Service to ROS 2 for topic '"

```

File: ros2/dynamic/src/Subscriber.cpp [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+    : participant_(participant)
+    , dds_subscriber_(nullptr)
+    , dynamic_data_(nullptr)
+    , topic_name_(topic_name)
+    , message_type_(message_type)
+    , is_callback_(is_callback)
+    , reception_threads_()
+    , stop_cleaner_(false)
+    , cleaner_thread_(&Subscriber::cleaner_function, this)
+    , logger_("is::sh::ROS2 Dynamic::Subscriber")
```
Comment (jamoralp) [inline ros2/dynamic/src/Subscriber.cpp]:
```suggestion
    , logger_("is::sh::ROS2::Dynamic::Subscriber")

```

File: ros2/dynamic/src/Subscriber.cpp [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+    }
+
+    dynamic_data_ = participant->create_dynamic_data(topic_name);
+
+    // Retrieve DDS participant
+    ::fastdds::dds::DomainParticipant* dds_participant = participant->get_dds_participant();
+    if (!dds_participant)
+    {
+        throw ROS2MiddlewareException(
+                  logger_, "Trying to create a subscriber without a DDS participant!");
```
Comment (jamoralp) [inline ros2/dynamic/src/Subscriber.cpp:89]:
```suggestion
                  logger_, "Trying to create a subscriber without a ROS 2 node!");

```

File: ros2/dynamic/src/Subscriber.cpp [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+    {
+        participant_->get_dds_participant()->delete_topic(dds_topic_);
+    }
+}
+
+void Subscriber::receive(
+        const fastrtps::types::DynamicData* dds_message)
+{
+    logger_ << utils::Logger::Level::INFO
+            << "Receiving message from DDS for topic '" << topic_name_ << "'" << std::endl;
```
Comment (jamoralp) [inline ros2/dynamic/src/Subscriber.cpp]:
```suggestion
            << "Receiving message from ROS 2 for topic '" << topic_name_ << "'" << std::endl;

```

File: ros2/dynamic/src/Subscriber.cpp [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+    {
+        logger_ << utils::Logger::Level::INFO
+                << "Received message: [[ " << is_message << " ]]" << std::endl;
+
+        (*is_callback_)(is_message);
+    }
+    else
+    {
+        logger_ << utils::Logger::Level::ERROR
+                << "Failed to convert message from DDS to Integration Service for topic '"
```
Comment (jamoralp) [inline ros2/dynamic/src/Subscriber.cpp]:
```suggestion
                << "Failed to convert message from ROS 2 to Integration Service for topic '"

```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 17]
```diff
@@ -0,0 +1,632 @@
...
+    bool configure(
+            const core::RequiredTypes& /*types*/,
+            const YAML::Node& configuration,
+            TypeRegistry& /*type_registry*/) override
+    {
+        /*
+         * The Fast-DDS sh doesn't define new types.
+         * Needed types will be defined in the 'types' section of the YAML file, and hence,
+         * already registered in the 'TypeRegistry' by the *Integration Service core*.
+         */
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]: This is not going to be true once we add the `using` param to the YAML file

File: ros2/dynamic/src/SystemHandle.cpp [hunk 18]
```diff
@@ -0,0 +1,632 @@
...
+
+    bool preprocess_types(
+        const YAML::Node& types_node) override
+    {
+        std::vector<std::string> include_paths;
+        if (types_node["paths"])
+        {
+            // Check if there are paths to custom idls
+            std::regex reg("/opt/ros/([a-z])+/share/*");
+            for (auto& path : types_node["paths"])
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
            for (const auto& path : types_node["paths"])
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 19]
```diff
@@ -0,0 +1,632 @@
...
+                    logger_ << utils::Logger::Level::DEBUG
+                            << "Added path for custom IDL." << std::endl;
+                }
+                include_paths.push_back(path.as<std::string>());
+            }
+        }
+
+        if(types_node["idls"])
+        {
+            for (auto& entry: types_node["idls"])
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
            for (const auto& entry : types_node["idls"])
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 20]
```diff
@@ -0,0 +1,632 @@
...
+                                << "Searching include clauses..." << std::endl;
+
+                while (iter != end)
+                {
+                    for(unsigned i = 0; i < iter->size(); ++i)
+                    {
+                        // On each include extracts the package_name
+                        std::string incl = (*iter)[i];
+                        logger_ << utils::Logger::Level::DEBUG
+                                    << "Include Clause: " << incl << std::endl;
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                << "Include Clause: " << incl << std::endl;
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 21]
```diff
@@ -0,0 +1,632 @@
...
+                        }
+
+                        if (custom_includes)
+                        {
+                            std::regex path_reg("[<\"][^>\"]*[\">]");
+                            std::cmatch cm1;
+                            std::regex_search(incl.c_str(), cm1, path_reg);
+                            std::string in_path = cm1[0];
+                            in_path = in_path.substr(1, in_path.length() - 2);
+                            for (auto& path: include_paths)
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                            for (const auto& path : include_paths)
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 22]
```diff
@@ -0,0 +1,632 @@
...
+                if (context.success)
+                {
+                    size_t modules = context.module().submodule_size() - ros2_modules.size();
+                    logger_ << utils::Logger::Level::DEBUG
+                            << "Number of modules: " << modules << std::endl;
+
+                    if (modules == 0)
+                    {
+                        logger_ << utils::Logger::Level::ERROR
+                                << "The type doesn't have module. Please follow the ROS2 naming convention."
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                << "The type is not declared within an IDL module. Please follow the ROS 2 naming convention."
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 23]
```diff
@@ -0,0 +1,632 @@
...
+                                << std::endl;
+
+                        return false;
+                    }
+                    else if (modules > 1)
+                    {
+                        logger_ << utils::Logger::Level::ERROR
+                                << "There can only be one module per idl, add"
+                                << " another entry in the YAML idls tag." << std::endl;
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                << "There can only be one module per IDL, add"
                                << " another entry in the YAML 'idls' tag." << std::endl;
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 24]
```diff
@@ -0,0 +1,632 @@
...
+                                            << std::endl;
+                                    is_success = false;
+                                    return;
+                                }
+
+                                // Check whether the following submodule is msg or srv
+                                if (!mod.has_submodule("msg") && !mod.has_submodule("srv"))
+                                {
+                                    logger_ << utils::Logger::Level::ERROR
+                                            << "The type must have a module msg or srv."
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                            << "The module '" << mod.name() << "' must be followed by a 'msg' or 'srv' submodule prior to declaring the type."
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 25]
```diff
@@ -0,0 +1,632 @@
...
+                                    logger_ << utils::Logger::Level::DEBUG
+                                        << "The file " << filename << " has been copied to " << path
+                                        << std::endl;
+                                    if (!std::filesystem::exists(path))
+                                    {
+                                        std::filesystem::copy(mv_idl, path);
+                                    }
+                                }
+
+                                std::string package_name = "--package_name " + mod.name();
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                const std::string package_name = "--package_name " + mod.name();
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 26]
```diff
@@ -0,0 +1,632 @@
...
+                                std::string path = "--install_path /opt/ros/" + ROS2_DISTRO;
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                const std::string path = "--install_path /opt/ros/" + ROS2_DISTRO;
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 27]
```diff
@@ -0,0 +1,632 @@
...
+                                    is_success = false;
+                                    return;
+                                }
+
+                                //TODO: Change when the services are implemented
+                                std::ofstream idlfile ("/tmp/" + mod.name() + "/msg/" + aux + ".idl");
+                                idlfile << idl << std::endl;
+                                idlfile.close();
+
+                                for (auto& mv_idl: custom_include_paths)
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                for (const auto& mv_idl : custom_include_paths)
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 28]
```diff
@@ -0,0 +1,632 @@
...
+                                {
+                                    std::string filename = std::filesystem::path(mv_idl).filename();
+                                    std::string path = std::filesystem::path(tmp / mod.name() / "msg" / filename);
+                                    logger_ << utils::Logger::Level::DEBUG
+                                        << "The file " << filename << " has been copied to " << path
+                                        << std::endl;
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                            << "The file " << filename << " has been copied to " << path
                                            << std::endl;
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 29]
```diff
@@ -0,0 +1,632 @@
...
+                                        << "Generating ROS2 Type Support for package: " << package_name
+                                        << std::endl;
+
+                                std::string command = "exec bash /tmp/generator.bash " + package_name + " " + path;
+
+                                if (ros2_modules.size() != 0)
+                                {
+                                    std::ostringstream stream;
+                                    std::copy(ros2_modules.begin(), ros2_modules.end(), std::ostream_iterator<std::string>(stream, ";"));
+                                    std::string depends = "--depedencies \"" + stream.str() + "\"";
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                    std::string depends = "--dependencies \"" + stream.str() + "\"";
```

Does this mean that the bash script needs to be changed too?

File: ros2/dynamic/src/SystemHandle.cpp [hunk 30]
```diff
@@ -0,0 +1,632 @@
...
+                                        << output << std::endl;
+
+                                int st = pclose(pipe);
+                                if (WIFEXITED(st))
+                                {
+                                    if (1 == WEXITSTATUS(st))
+                                    {
+                                        logger_ << is::utils::Logger::Level::ERROR
+                                                << "Failed to generate the Type Support for package " << package_name
+                                                << ". Make sure you follow all the ROS2 naming and structure convention"
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                                << "Failed to generate the Type Support for package '" << package_name
                                                << "'. Make sure you follow all the ROS2 naming and structure convention"
```

File: ros2/dynamic/src/SystemHandle.cpp [hunk 31]
```diff
@@ -0,0 +1,632 @@
...
+                                                    << output << std::endl;
+                                        }
+
+                                        is_success = false;
+                                        return;
+                                    }
+                                    else if (0 == WEXITSTATUS(st))
+                                    {
+                                        logger_ << is::utils::Logger::Level::INFO
+                                                << "ROS2 Type Supports generation finished. Installed in path "
```
Comment (jamoralp) [inline ros2/dynamic/src/SystemHandle.cpp]:
```suggestion
                                                << "ROS2 Type Supports generation finished for package '" << package_name << "', installed in path "
```

File: ros2/dynamic/CMakeLists.txt (hunk 3 above)
Comment (lauramg15) [reply ros2/dynamic/CMakeLists.txt]: No

**Review by jamoralp — DISMISSED** [review]

**Pull request closed without merging**